src/sinks/humio_logs.rs

```diff
@@ -118,3 +118,156 @@ mod tests {
+            );
+        });
+    }
+
+    #[derive(Debug)]
+    struct HumioRepository {
+        name: String,
+        default_ingest_token: String,
+    }
+
+    /// create a new test config with the given ingest token
+    fn config(token: &str) -> super::HumioLogsConfig {
+        super::HumioLogsConfig {
+            host: Some(HOST.to_string()),
+            token: token.to_string(),
+            compression: Compression::None,
+            encoding: Encoding::Json.into(),
+            batch: BatchConfig {
+                max_events: Some(1),
+                ..Default::default()
+            },
+            ..Default::default()
+        }
+    }
+
+    /// create a new test humio repository to publish to
+    async fn create_repository() -> HumioRepository {
+        let client = reqwest::Client::builder().build().unwrap();
+
+        // https://docs.humio.com/api/graphql/
+        let graphql_url = format!("{}/graphql", HOST);
+
+        let name = random_string(50);
+
+        let params = json!({
+        "query": format!(
+            r#"
+mutation {{
+  createRepository(name:"{}") {{
+    repository {{
+      name
+      type
+      ingestTokens {{
+        name
+        token
+      }}
+    }}
+  }}
+}}
+"#,
+            name
+        ),
+        });
+
+        let res = client
+            .post(&graphql_url)
+            .json(&params)
+            .send()
+            .await
+            .unwrap();
+
+        let json: JsonValue = res.json().await.unwrap();
+        let repository = &json["data"]["createRepository"]["repository"];
+
+        let token = repository["ingestTokens"].as_array().unwrap()[0]["token"]
+            .as_str()
+            .unwrap()
+            .to_string();
+
+        HumioRepository {
+            name: repository["name"].as_str().unwrap().to_string(),
+            default_ingest_token: token,
+        }
+    }
+
+    /// fetch event from the repository that has a matching message value
+    async fn find_entry(repository_name: &str, message: &str) -> serde_json::value::Value {
+        let client = reqwest::Client::builder().build().unwrap();
+
+        // https://docs.humio.com/api/using-the-search-api-with-humio
+        let search_url = format!("{}/api/v1/repositories/{}/query", HOST, repository_name);
+        let search_query = format!(r#"message="{}""#, message);
+
+        // poll multiple times for event to show up
+        for _ in 0..20usize {
+            let res = client
+                .post(&search_url)
+                .json(&json!({
+                    "queryString": search_query,
+                }))
+                .header(reqwest::header::ACCEPT, "application/json")
+                .send()
+                .await
+                .unwrap();
+
+            let json: JsonValue = res.json().await.unwrap();
+
+            let results = json.as_array().unwrap();
+            if !results.is_empty() {
+                return results[0].clone();
```

Comment:
Is the clone needed here?

Reply:
Without it, it complains that I cannot move out of the index:

```
error[E0507]: cannot move out of index of `std::vec::Vec<serde_json::value::Value>`
   --> src/sinks/humio_logs.rs:266:24
    |
266 |                 return results[0];
    |                        ^^^^^^^^^^ move occurs because value has type `serde_json::value::Value`, which does not implement the `Copy` trait

```

I definitely welcome thoughts on how I could avoid cloning here though.

Reply:
Ah, yes, I was thinking it was an array, in which case Rust would be able to reason more about ownership.